File: form_validation.py
"""
Enhanced Form Validation and Input Handling for CodeXam Platform
Provides comprehensive client-side and server-side validation with security features
"""

import re
import html
from typing import Dict, Any, List, Optional, Tuple
from flask import request, jsonify
import logging
from datetime import datetime, timedelta
import hashlib
import secrets
from functools import wraps

logger = logging.getLogger(__name__)


class ValidationError(Exception):
    """Custom exception for validation errors."""
    pass


class SecurityViolation(Exception):
    """Custom exception for security violations."""
    pass


class RateLimiter:
    """Simple in-memory rate limiter."""
    
    def __init__(self):
        self.requests = {}
        self.blocked_ips = {}
    
    def is_rate_limited(self, ip: str, limit: int = 10, window: int = 60) -> bool:
        """Check if IP is rate limited."""
        now = datetime.now()
        
        # Clean old requests
        self._cleanup_old_requests(now, window)
        
        # Check if IP is temporarily blocked
        if ip in self.blocked_ips:
            if now < self.blocked_ips[ip]:
                return True
            else:
                del self.blocked_ips[ip]
        
        # Count requests in window
        if ip not in self.requests:
            self.requests[ip] = []
        
        self.requests[ip] = [req_time for req_time in self.requests[ip] 
                           if now - req_time < timedelta(seconds=window)]
        
        if len(self.requests[ip]) >= limit:
            # Block IP for extended period
            self.blocked_ips[ip] = now + timedelta(minutes=5)
            logger.warning(f"IP {ip} blocked for rate limiting")
            return True
        
        self.requests[ip].append(now)
        return False
    
    def _cleanup_old_requests(self, now: datetime, window: int):
        """Clean up old request records."""
        cutoff = now - timedelta(seconds=window * 2)
        for ip in list(self.requests.keys()):
            self.requests[ip] = [req_time for req_time in self.requests[ip] 
                               if req_time > cutoff]
            if not self.requests[ip]:
                del self.requests[ip]


class CSRFProtection:
    """CSRF protection for forms."""
    
    def __init__(self):
        self.tokens = {}
    
    def generate_token(self, session_id: str) -> str:
        """Generate CSRF token for session."""
        token = secrets.token_urlsafe(32)
        self.tokens[session_id] = {
            'token': token,
            'created': datetime.now()
        }
        return token
    
    def validate_token(self, session_id: str, token: str) -> bool:
        """Validate CSRF token."""
        if session_id not in self.tokens:
            return False
        
        stored_token = self.tokens[session_id]['token']
        created = self.tokens[session_id]['created']
        
        # Token expires after 1 hour
        if datetime.now() - created > timedelta(hours=1):
            del self.tokens[session_id]
            return False
        
        return stored_token == token
    
    def cleanup_expired_tokens(self):
        """Clean up expired tokens."""
        now = datetime.now()
        expired = [sid for sid, data in self.tokens.items() 
                  if now - data['created'] > timedelta(hours=1)]
        for sid in expired:
            del self.tokens[sid]


# Global instances
rate_limiter = RateLimiter()
csrf_protection = CSRFProtection()


class InputValidator:
    """Comprehensive input validation system."""
    
    # Security patterns to detect
    MALICIOUS_PATTERNS = [
        r'<script[^>]*>.*?</script>',  # Script tags
        r'javascript:',                # JavaScript protocol
        r'on\w+\s*=',                 # Event handlers
        r'eval\s*\(',                 # eval() calls
        r'expression\s*\(',           # CSS expressions
        r'vbscript:',                 # VBScript protocol
        r'<iframe[^>]*>',             # iframe tags
        r'<object[^>]*>',             # object tags
        r'<embed[^>]*>',              # embed tags
        r'<link[^>]*>',               # link tags
        r'<meta[^>]*>',               # meta tags
        r'<form[^>]*>',               # form tags
    ]
    
    # SQL injection patterns
    SQL_INJECTION_PATTERNS = [
        r'(union|select|insert|update|delete|drop|create|alter)\s+',
        r'(or|and)\s+\d+\s*=\s*\d+',
        r'[\'\"]\s*(or|and)\s+[\'\"]\d+[\'\"]\s*=\s*[\'\"]\d+[\'\"]*',
        r';\s*(drop|delete|update|insert)',
        r'exec\s*\(',
        r'sp_\w+',
    ]
    
    @staticmethod
    def sanitize_string(value: str, max_length: int = 1000) -> str:
        """Sanitize string input."""
        if not isinstance(value, str):
            raise ValidationError("Input must be a string")
        
        # Length check
        if len(value) > max_length:
            raise ValidationError(f"Input exceeds maximum length of {max_length} characters")
        
        # HTML escape
        sanitized = html.escape(value.strip())
        
        # Check for malicious patterns
        for pattern in InputValidator.MALICIOUS_PATTERNS:
            if re.search(pattern, sanitized, re.IGNORECASE):
                raise SecurityViolation(f"Potentially malicious input detected")
        
        # Check for SQL injection
        for pattern in InputValidator.SQL_INJECTION_PATTERNS:
            if re.search(pattern, sanitized, re.IGNORECASE):
                raise SecurityViolation("Potential SQL injection detected")
        
        return sanitized
    
    @staticmethod
    def validate_problem_id(problem_id: Any) -> int:
        """Validate problem ID."""
        try:
            pid = int(problem_id)
            if pid <= 0:
                raise ValidationError("Problem ID must be a positive integer")
            if pid > 1000000:  # Reasonable upper limit
                raise ValidationError("Problem ID is too large")
            return pid
        except (ValueError, TypeError):
            raise ValidationError("Problem ID must be a valid integer")
    
    @staticmethod
    def validate_language(language: str, supported_languages: List[str]) -> str:
        """Validate programming language."""
        if not language or not isinstance(language, str):
            raise ValidationError("Programming language is required")
        
        language = language.lower().strip()
        
        if language not in supported_languages:
            raise ValidationError(
                f"Unsupported language: {language}. "
                f"Supported languages: {', '.join(supported_languages)}"
            )
        
        return language
    
    @staticmethod
    def validate_code(code: str, language: str, max_length: int = 50000) -> str:
        """Validate code input."""
        if not code or not isinstance(code, str):
            raise ValidationError("Code cannot be empty")
        
        code = code.strip()
        
        if not code:
            raise ValidationError("Code cannot be empty")
        
        if len(code) > max_length:
            raise ValidationError(
                f"Code exceeds maximum length of {max_length} characters. "
                f"Current length: {len(code)}"
            )
        
        # Language-specific validation
        if language == 'python':
            InputValidator._validate_python_code(code)
        elif language == 'javascript':
            InputValidator._validate_javascript_code(code)
        
        return code
    
    @staticmethod
    def _validate_python_code(code: str):
        """Python-specific code validation."""
        # Check for dangerous imports
        dangerous_imports = [
            'os', 'sys', 'subprocess', 'socket', 'urllib', 'requests',
            'http', 'ftplib', 'smtplib', 'webbrowser', 'tempfile',
            'shutil', 'glob', 'pickle', 'marshal'
        ]
        
        lines = code.split('\n')
        for line_num, line in enumerate(lines, 1):
            line = line.strip().lower()
            
            # Check for dangerous imports
            for dangerous in dangerous_imports:
                if f'import {dangerous}' in line or f'from {dangerous}' in line:
                    raise SecurityViolation(
                        f"Restricted import '{dangerous}' found on line {line_num}. "
                        "File operations and network access are not allowed."
                    )
            
            # Check for dangerous functions
            dangerous_functions = ['open(', 'file(', 'input(', 'eval(', 'exec(']
            for func in dangerous_functions:
                if func in line:
                    raise SecurityViolation(
                        f"Restricted function '{func.rstrip('(')}' found on line {line_num}. "
                        "This function is not allowed for security reasons."
                    )
    
    @staticmethod
    def _validate_javascript_code(code: str):
        """JavaScript-specific code validation."""
        dangerous_patterns = [
            'require(', 'import(', 'eval(', 'process.', 'global.',
            '__dirname', '__filename', 'fs.', 'child_process',
            'http.', 'https.', 'net.', 'crypto.'
        ]
        
        lines = code.split('\n')
        for line_num, line in enumerate(lines, 1):
            line = line.strip().lower()
            
            for pattern in dangerous_patterns:
                if pattern in line:
                    raise SecurityViolation(
                        f"Restricted operation '{pattern.rstrip('.')}' found on line {line_num}. "
                        "File operations and network access are not allowed."
                    )
    
    @staticmethod
    def validate_user_name(name: str) -> str:
        """Validate user name input."""
        if not name or not isinstance(name, str):
            raise ValidationError("Name is required")
        
        name = name.strip()
        
        if not name:
            raise ValidationError("Name cannot be empty")
        
        if len(name) > 50:
            raise ValidationError("Name cannot exceed 50 characters")
        
        if len(name) < 2:
            raise ValidationError("Name must be at least 2 characters long")
        
        # Allow only alphanumeric, spaces, and common punctuation
        if not re.match(r'^[a-zA-Z0-9\s\-_.]+$', name):
            raise ValidationError(
                "Name can only contain letters, numbers, spaces, hyphens, underscores, and periods"
            )
        
        return InputValidator.sanitize_string(name, 50)
    
    @staticmethod
    def validate_admin_input(data: Dict[str, Any]) -> Dict[str, Any]:
        """Validate admin form inputs."""
        validated = {}
        
        # Problem title
        if 'title' in data:
            title = InputValidator.sanitize_string(data['title'], 200)
            if len(title) < 3:
                raise ValidationError("Problem title must be at least 3 characters long")
            validated['title'] = title
        
        # Problem description
        if 'description' in data:
            description = InputValidator.sanitize_string(data['description'], 5000)
            if len(description) < 10:
                raise ValidationError("Problem description must be at least 10 characters long")
            validated['description'] = description
        
        # Difficulty
        if 'difficulty' in data:
            difficulty = data['difficulty'].strip().title()
            if difficulty not in ['Easy', 'Medium', 'Hard']:
                raise ValidationError("Difficulty must be Easy, Medium, or Hard")
            validated['difficulty'] = difficulty
        
        return validated


def rate_limit(limit: int = 10, window: int = 60):
    """Decorator for rate limiting."""
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            client_ip = request.environ.get('HTTP_X_FORWARDED_FOR', request.remote_addr)
            
            if rate_limiter.is_rate_limited(client_ip, limit, window):
                return jsonify({
                    "status": "ERROR",
                    "error": {
                        "type": "RATE_LIMIT_EXCEEDED",
                        "message": "Too many requests. Please try again later.",
                        "retry_after": 300  # 5 minutes
                    }
                }), 429
            
            return f(*args, **kwargs)
        return decorated_function
    return decorator


def csrf_protect(f):
    """Decorator for CSRF protection."""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if request.method == 'POST':
            session_id = request.environ.get('HTTP_X_SESSION_ID', 'default')
            token = request.form.get('csrf_token') or request.headers.get('X-CSRF-Token')
            
            if not token or not csrf_protection.validate_token(session_id, token):
                return jsonify({
                    "status": "ERROR",
                    "error": {
                        "type": "CSRF_VIOLATION",
                        "message": "Invalid or missing CSRF token",
                    }
                }), 403
        
        return f(*args, **kwargs)
    return decorated_function


def validate_submission_request(request_data: Dict[str, Any]) -> Tuple[int, str, str]:
    """Validate code submission request with enhanced security."""
    try:
        # Extract and validate problem ID
        problem_id = InputValidator.validate_problem_id(request_data.get('problem_id'))
        
        # Extract and validate language
        supported_languages = ['python', 'javascript', 'java', 'cpp']
        language = InputValidator.validate_language(
            request_data.get('language', ''), 
            supported_languages
        )
        
        # Extract and validate code
        code = InputValidator.validate_code(
            request_data.get('code', ''), 
            language
        )
        
        return problem_id, code, language
        
    except (ValidationError, SecurityViolation) as e:
        raise ValidationError(str(e))
    except Exception as e:
        logger.error(f"Unexpected validation error: {e}")
        raise ValidationError("Invalid submission data")


def validate_user_name_request(name: str) -> str:
    """Validate user name setting request."""
    try:
        return InputValidator.validate_user_name(name)
    except (ValidationError, SecurityViolation) as e:
        raise ValidationError(str(e))
    except Exception as e:
        logger.error(f"Unexpected name validation error: {e}")
        raise ValidationError("Invalid name")


def validate_admin_request(request_data: Dict[str, Any]) -> Dict[str, Any]:
    """Validate admin form request."""
    try:
        return InputValidator.validate_admin_input(request_data)
    except (ValidationError, SecurityViolation) as e:
        raise ValidationError(str(e))
    except Exception as e:
        logger.error(f"Unexpected admin validation error: {e}")
        raise ValidationError("Invalid admin input")


def get_client_info(request) -> Dict[str, str]:
    """Get client information for logging and security."""
    return {
        'ip': request.environ.get('HTTP_X_FORWARDED_FOR', request.remote_addr),
        'user_agent': request.headers.get('User-Agent', 'Unknown'),
        'referer': request.headers.get('Referer', 'Unknown'),
        'timestamp': datetime.now().isoformat()
    }


def log_security_event(event_type: str, details: Dict[str, Any], request=None):
    """Log security events for monitoring."""
    log_data = {
        'event_type': event_type,
        'details': details,
        'timestamp': datetime.now().isoformat()
    }
    
    if request:
        log_data['client_info'] = get_client_info(request)
    
    logger.warning(f"Security event: {event_type} - {details}")


# Periodic cleanup function
def cleanup_security_data():
    """Clean up expired security data."""
    csrf_protection.cleanup_expired_tokens()
    # Rate limiter cleanup is handled automatically

File: static/js/form-validation.js
/**
 * Enhanced Client-Side Form Validation for CodeXam Platform
 * Provides real-time validation and user feedback
 */

class CodeXamValidator {
    constructor() {
        this.initialized = false;
        this.validationRules = {
            code: {
                maxLength: 50000,
                minLength: 1,
                required: true
            },
            language: {
                required: true,
                allowedValues: ['python', 'javascript', 'java', 'cpp']
            },
            problemId: {
                required: true,
                type: 'number',
                min: 1
            },
            userName: {
                maxLength: 50,
                minLength: 2,
                pattern: /^[a-zA-Z0-9\s\-_.]+$/,
                required: true
            }
        };
        
        this.errorMessages = {
            required: 'This field is required',
            maxLength: 'Input exceeds maximum allowed length',
            minLength: 'Input is too short',
            pattern: 'Invalid characters detected',
            invalidLanguage: 'Please select a supported programming language',
            invalidProblemId: 'Invalid problem ID',
            emptyCode: 'Please enter your code solution',
            codeTooLong: 'Code exceeds maximum length of 50,000 characters',
            securityViolation: 'Code contains restricted operations',
            maliciousContent: 'Potentially unsafe content detected'
        };
        
        this.securityPatterns = [
            /import\s+(os|sys|subprocess|socket|urllib|requests|http|ftplib|smtplib|webbrowser|tempfile|shutil|glob|pickle|marshal)/gi,
            /from\s+(os|sys|subprocess|socket|urllib|requests|http|ftplib|smtplib|webbrowser|tempfile|shutil|glob|pickle|marshal)/gi,
            /(open|file|input|eval|exec)\s*\(/gi,
            /require\s*\(['"][^'"]*fs['"]\)/gi,
            /require\s*\(['"][^'"]*child_process['"]\)/gi,
            /process\./gi,
            /__dirname|__filename/gi,
            /<script[^>]*>.*?<\/script>/gi,
            /javascript:/gi,
            /on\w+\s*=/gi
        ];
        
        this.init();
    }
    
    init() {
        if (this.initialized) return;
        
        // Initialize validation on form elements
        this.setupFormValidation();
        this.setupRealTimeValidation();
        this.setupSecurityChecks();
        
        this.initialized = true;
        console.log('CodeXam Validator initialized');
    }
    
    setupFormValidation() {
        // Main submission form
        const submitForm = document.querySelector('#submission-form, .submission-form');
        if (submitForm) {
            submitForm.addEventListener('submit', (e) => {
                if (!this.validateSubmissionForm(submitForm)) {
                    e.preventDefault();
                }
            });
        }
        
        // User name form
        const nameForm = document.querySelector('#name-form, .name-form');
        if (nameForm) {
            nameForm.addEventListener('submit', (e) => {
                if (!this.validateNameForm(nameForm)) {
                    e.preventDefault();
                }
            });
        }
        
        // Admin forms
        const adminForms = document.querySelectorAll('.admin-form');
        adminForms.forEach(form => {
            form.addEventListener('submit', (e) => {
                if (!this.validateAdminForm(form)) {
                    e.preventDefault();
                }
            });
        });
    }
    
    setupRealTimeValidation() {
        // Code editor validation
        const codeEditor = document.querySelector('#code-editor, .code-editor, textarea[name="code"]');
        if (codeEditor) {
            let debounceTimer;
            codeEditor.addEventListener('input', (e) => {
                clearTimeout(debounceTimer);
                debounceTimer = setTimeout(() => {
                    this.validateCodeInput(codeEditor);
                }, 300);
            });
            
            codeEditor.addEventListener('blur', () => {
                this.validateCodeInput(codeEditor);
            });
        }
        
        // Language selection validation
        const languageSelect = document.querySelector('select[name="language"], #language-select');
        if (languageSelect) {
            languageSelect.addEventListener('change', (e) => {
                this.validateLanguageSelection(languageSelect);
                this.updateCodeTemplate(languageSelect.value);
            });
        }
        
        // User name validation
        const nameInput = document.querySelector('input[name="user_name"], #user-name');
        if (nameInput) {
            nameInput.addEventListener('input', (e) => {
                this.validateUserNameInput(nameInput);
            });
        }
    }
    
    setupSecurityChecks() {
        // Monitor for potential security issues
        document.addEventListener('paste', (e) => {
            const target = e.target;
            if (target.matches('textarea[name="code"], .code-editor')) {
                setTimeout(() => {
                    this.performSecurityCheck(target.value);
                }, 100);
            }
        });
    }
    
    validateSubmissionForm(form) {
        const data = new FormData(form);
        const code = data.get('code') || '';
        const language = data.get('language') || '';
        const problemId = data.get('problem_id') || '';
        
        let isValid = true;
        this.clearErrors(form);
        
        // Validate problem ID
        if (!this.validateProblemId(problemId)) {
            this.showError(form, 'problem_id', this.errorMessages.invalidProblemId);
            isValid = false;
        }
        
        // Validate language
        if (!this.validateLanguage(language)) {
            this.showError(form, 'language', this.errorMessages.invalidLanguage);
            isValid = false;
        }
        
        // Validate code
        const codeValidation = this.validateCode(code, language);
        if (!codeValidation.valid) {
            this.showError(form, 'code', codeValidation.message);
            isValid = false;
        }
        
        // Security check
        const securityCheck = this.performSecurityCheck(code);
        if (!securityCheck.safe) {
            this.showError(form, 'code', securityCheck.message);
            isValid = false;
        }
        
        return isValid;
    }
    
    validateNameForm(form) {
        const data = new FormData(form);
        const name = data.get('user_name') || data.get('name') || '';
        
        let isValid = true;
        this.clearErrors(form);
        
        const nameValidation = this.validateUserName(name);
        if (!nameValidation.valid) {
            this.showError(form, 'user_name', nameValidation.message);
            isValid = false;
        }
        
        return isValid;
    }
    
    validateAdminForm(form) {
        let isValid = true;
        this.clearErrors(form);
        
        // Validate title
        const title = form.querySelector('input[name="title"]');
        if (title && !this.validateTitle(title.value)) {
            this.showError(form, 'title', 'Title must be at least 3 characters long');
            isValid = false;
        }
        
        // Validate description
        const description = form.querySelector('textarea[name="description"]');
        if (description && !this.validateDescription(description.value)) {
            this.showError(form, 'description', 'Description must be at least 10 characters long');
            isValid = false;
        }
        
        return isValid;
    }
    
    validateCode(code, language) {
        if (!code || code.trim().length === 0) {
            return { valid: false, message: this.errorMessages.emptyCode };
        }
        
        if (code.length > this.validationRules.code.maxLength) {
            return { 
                valid: false, 
                message: `Code exceeds maximum length of ${this.validationRules.code.maxLength.toLocaleString()} characters (current: ${code.length.toLocaleString()})` 
            };
        }
        
        // Language-specific validation
        if (language === 'python') {
            const pythonValidation = this.validatePythonCode(code);
            if (!pythonValidation.valid) return pythonValidation;
        } else if (language === 'javascript') {
            const jsValidation = this.validateJavaScriptCode(code);
            if (!jsValidation.valid) return jsValidation;
        }
        
        return { valid: true };
    }
    
    validatePythonCode(code) {
        // Check for dangerous imports
        const dangerousImports = [
            'import os', 'import sys', 'import subprocess', 'import socket',
            'import urllib', 'import requests', 'import tempfile', 'import shutil',
            'from os', 'from sys', 'from subprocess', 'from socket'
        ];
        
        for (const dangerous of dangerousImports) {
            if (code.toLowerCase().includes(dangerous.toLowerCase())) {
                return {
                    valid: false,
                    message: `Restricted import detected: ${dangerous.split(' ')[1]}. File operations and network access are not allowed.`
                };
            }
        }
        
        // Check for dangerous functions
        const dangerousFunctions = ['open(', 'file(', 'input(', 'eval(', 'exec('];
        for (const func of dangerousFunctions) {
            if (code.includes(func)) {
                return {
                    valid: false,
                    message: `Restricted function detected: ${func.replace('(', '')}. This function is not allowed for security reasons.`
                };
            }
        }
        
        return { valid: true };
    }
    
    validateJavaScriptCode(code) {
        const dangerousPatterns = [
            'require(', 'import(', 'eval(', 'process.', 'global.',
            '__dirname', '__filename', 'fs.', 'child_process'
        ];
        
        for (const pattern of dangerousPatterns) {
            if (code.includes(pattern)) {
                return {
                    valid: false,
                    message: `Restricted operation detected: ${pattern.replace('.', '')}. File operations and network access are not allowed.`
                };
            }
        }
        
        return { valid: true };
    }
    
    validateLanguage(language) {
        return this.validationRules.language.allowedValues.includes(language);
    }
    
    validateProblemId(problemId) {
        const id = parseInt(problemId);
        return !isNaN(id) && id > 0;
    }
    
    validateUserName(name) {
        if (!name || name.trim().length === 0) {
            return { valid: false, message: this.errorMessages.required };
        }
        
        if (name.length < this.validationRules.userName.minLength) {
            return { valid: false, message: `Name must be at least ${this.validationRules.userName.minLength} characters long` };
        }
        
        if (name.length > this.validationRules.userName.maxLength) {
            return { valid: false, message: `Name cannot exceed ${this.validationRules.userName.maxLength} characters` };
        }
        
        if (!this.validationRules.userName.pattern.test(name)) {
            return { valid: false, message: 'Name can only contain letters, numbers, spaces, hyphens, underscores, and periods' };
        }
        
        return { valid: true };
    }
    
    validateTitle(title) {
        return title && title.trim().length >= 3;
    }
    
    validateDescription(description) {
        return description && description.trim().length >= 10;
    }
    
    performSecurityCheck(content) {
        for (const pattern of this.securityPatterns) {
            if (pattern.test(content)) {
                return {
                    safe: false,
                    message: 'Code contains restricted operations that are not allowed for security reasons.'
                };
            }
        }
        
        return { safe: true };
    }
    
    validateCodeInput(element) {
        const code = element.value;
        const language = this.getCurrentLanguage();
        
        // Update character count
        this.updateCharacterCount(element, code);
        
        // Validate code
        const validation = this.validateCode(code, language);
        if (!validation.valid) {
            this.showFieldError(element, validation.message);
        } else {
            this.clearFieldError(element);
        }
        
        // Security check
        const securityCheck = this.performSecurityCheck(code);
        if (!securityCheck.safe) {
            this.showFieldWarning(element, securityCheck.message);
        }
    }
    
    validateLanguageSelection(element) {
        const language = element.value;
        
        if (!this.validateLanguage(language)) {
            this.showFieldError(element, this.errorMessages.invalidLanguage);
        } else {
            this.clearFieldError(element);
        }
    }
    
    validateUserNameInput(element) {
        const name = element.value;
        const validation = this.validateUserName(name);
        
        if (!validation.valid) {
            this.showFieldError(element, validation.message);
        } else {
            this.clearFieldError(element);
        }
    }
    
    getCurrentLanguage() {
        const languageSelect = document.querySelector('select[name="language"], #language-select');
        return languageSelect ? languageSelect.value : 'python';
    }
    
    updateCharacterCount(element, text) {
        const counter = element.parentElement.querySelector('.character-count');
        if (counter) {
            const maxLength = this.validationRules.code.maxLength;
            const currentLength = text.length;
            const percentage = (currentLength / maxLength) * 100;
            
            counter.textContent = `${currentLength.toLocaleString()} / ${maxLength.toLocaleString()} characters`;
            
            if (percentage > 90) {
                counter.className = 'character-count text-danger';
            } else if (percentage > 75) {
                counter.className = 'character-count text-warning';
            } else {
                counter.className = 'character-count text-muted';
            }
        }
    }
    
    updateCodeTemplate(language) {
        const codeEditor = document.querySelector('#code-editor, .code-editor, textarea[name="code"]');
        if (!codeEditor || codeEditor.value.trim().length > 0) return;
        
        const templates = {
            python: 'def solution(args):\n    # Your code here\n    pass',
            javascript: 'function solution(args) {\n    // Your code here\n    return null;\n}',
            java: 'public class Solution {\n    public static Object solution(Object args) {\n        // Your code here\n        return null;\n    }\n}',
            cpp: '#include <iostream>\n\nusing namespace std;\n\nint solution(int args) {\n    // Your code here\n    return 0;\n}'
        };
        
        if (templates[language]) {
            codeEditor.value = templates[language];
            this.updateCharacterCount(codeEditor, templates[language]);
        }
    }
    
    showError(form, fieldName, message) {
        const field = form.querySelector(`[name="${fieldName}"]`);
        if (field) {
            this.showFieldError(field, message);
        }
    }
    
    showFieldError(field, message) {
        this.clearFieldError(field);
        
        field.classList.add('is-invalid');
        
        const errorDiv = document.createElement('div');
        errorDiv.className = 'invalid-feedback';
        errorDiv.textContent = message;
        
        field.parentElement.appendChild(errorDiv);
    }
    
    showFieldWarning(field, message) {
        const existingWarning = field.parentElement.querySelector('.text-warning');
        if (existingWarning) return;
        
        const warningDiv = document.createElement('div');
        warningDiv.className = 'text-warning small mt-1';
        warningDiv.innerHTML = `<i class="fas fa-exclamation-triangle"></i> ${message}`;
        
        field.parentElement.appendChild(warningDiv);
    }
    
    clearFieldError(field) {
        field.classList.remove('is-invalid');
        const errorFeedback = field.parentElement.querySelector('.invalid-feedback');
        if (errorFeedback) {
            errorFeedback.remove();
        }
        
        const warning = field.parentElement.querySelector('.text-warning');
        if (warning) {
            warning.remove();
        }
    }
    
    clearErrors(form) {
        const invalidFields = form.querySelectorAll('.is-invalid');
        invalidFields.forEach(field => {
            this.clearFieldError(field);
        });
    }
}

// Initialize validator when DOM is ready
document.addEventListener('DOMContentLoaded', () => {
    window.codexamValidator = new CodeXamValidator();
});

// Export for testing
if (typeof module !== 'undefined' && module.exports) {
    module.exports = CodeXamValidator;
}
